### DISCUSSION TITLE
Error while running MNIST pre-built finn example

### DISCUSSION THREAD

**Discussion by negative-slack** [Xilinx/finn#785] [Q&A] [answered]
Hello everyone, 

I am trying to run the pre-built finn example, but I am facing this error. Where to find this bitfile ? 

![error0](https://user-images.githubusercontent.com/97265693/225113525-c10c563c-7533-4931-954f-3579f5536fee.png)

**Comment by mmrahorovic** [comment, marked as answer]
Hi @n17a10d19e97r , thanks for your question!

From the latest FINN-examples release, we currently don't provide pre-built bitfiles for the Pynq-Z2 board. If you want to generate a bitfile for the TFC networks for the Pynq-Z2 board, you would have to do that manually. You can follow the steps [here](https://github.com/Xilinx/finn-examples/tree/main/build/bnn-pynq). More specifically, you would only need to edit the [following line](https://github.com/Xilinx/finn-examples/blob/3cdec2b90c598b4a26e9a6a3c5510bd217b64cf6/build/bnn-pynq/build.py#L20) to `zynq_platforms = ["Pynq-Z2"]` and `alveo_platforms=[]` and then run the `bnn-pynq/build.py` build flow.
Alternatively, you could use a Pynq-Z1 board for which we have included pre-built bitfiles. 

Hope this helps you further. Please let me know if you have any further questions!

**Reply by JJJWG** [reply]
Hi @mmrahorovic ,

I followed your guide to build the bitfiles manually for the Pynq-Z2 board as you described. After successfully building the bitfile for the Cybersecurity-FINN-Example I got the following output with the directories as shown in the Image:
`~/finn-examples/build/cybersecurity-mlp/output_unsw_nb15-mlp-w2a2_Pynq-Z2`:
![image](https://user-images.githubusercontent.com/122390554/235764025-bdc12cab-71d6-4d73-8d13-7224ceec1522.png)
![image](https://user-images.githubusercontent.com/122390554/235764133-9cda6eaf-f24f-4bc1-a476-fbcafc06c170.png)
 
How do I proceed from here and what should I do with the re-build bitfiles or the other output-files so that I can use the provided jupyter notebooks from FINN-Examples on my board? I am asking because I still have the same error as shown above, even tough I re-build the bitfile.

**Reply by mmrahorovic** [reply]
Hi @JJJWG,

Good question! The generated output files/folders are looking good from what I can tell.
In order to test the generated bitfile with one of the notebooks, you will have to make some changes to the notebook. At the moment, there is not a 'clean way' of doing this, so I will highlight the steps:
- In each notebook, a `FINNExampleOverlay` is created, which inherits the `Pynq Overlay` class and adds a few FINN-specific methods. When you create the [FINNExampleOverlay](https://github.com/Xilinx/finn-examples/blob/c75c1168b84a3e40b6ce68fc8ac6aef99fab81f3/finn_examples/driver.py#L47), among others, a bitfile is downloaded on the FPGA. This is happening usually in the second code-block of each notebook. For the cybersecurity-MLP example: `accel = models.mlp_w2a2_unsw_nb15()`.
- Looking at that [function](https://github.com/Xilinx/finn-examples/blob/c75c1168b84a3e40b6ce68fc8ac6aef99fab81f3/finn_examples/models.py#L349), you will need to change the `filename` variable to point to your newly generated bitfile. To do so, copy over the `bitfile` folder that was generated in the build process to your target board. Then, replace `accel = models.mlp_w2a2_unsw_nb15()` to the following:

```Python
from finn_examples.models import resolve_target_platform, get_driver_mode
from finn_examples.driver import FINNExampleOverlay
from qonnx.core.datatype import DataType

unsw_nb15_mlp_io_shape_dict = {
    "idt" : [DataType['BIPOLAR']],
    "odt" : [DataType['BIPOLAR']],
    "ishape_normal" : [(1, 600)],
    "oshape_normal" : [(1, 1)],
    "ishape_folded" : [(1, 15, 40)],
    "oshape_folded" : [(1, 1, 1)],
    "ishape_packed" : [(1, 15, 5)],
    "oshape_packed" : [(1, 1, 1)],
    "input_dma_name" : ['idma0'],
    "output_dma_name" : ['odma0'],
    "number_of_external_weights": 0,
    "num_inputs" : 1,
    "num_outputs" : 1,
}

def mlp_w2a2_unsw_nb15(target_platform=None):
    target_platform = resolve_target_platform(target_platform)
    driver_mode = get_driver_mode()
    filename = "/home/xilinx/bitfile/finn-accel.bit" # Path to the finn-accel.{bit,hwh} -- replace this with the location you copy the bitfile directory to
    fclk_mhz = 100.0
    return FINNExampleOverlay(
        filename,
        driver_mode,
        unsw_nb15_mlp_io_shape_dict,
        fclk_mhz=fclk_mhz
    )

accel = mlp_w2a2_unsw_nb15()
```

- You should now be good to go to run the rest of the [notebook](https://github.com/Xilinx/finn-examples/blob/main/finn_examples/notebooks/6_cybersecurity_with_mlp.ipynb) cells.

I have not tested this on my end, so please let me know if you run into any other problems!
